### PR TITLE
fix(gatsby): `@link` directives with arrays were broken

### PR DESCRIPTION
**Pull request by pvdz**
A schema can use `@link` to connect certain parts together implicitly. The resolver that deals with `@link` was ignoring the field type and always compiling an `$eq` operator.

Using `$eq` on arrays worked fine in Sift but when we tossed Sift, this broke (2.23.0). There are no real tests around `@link` and it's an edge case in a not commonly used feature, so we haven't seen many reports of problems related to this.

The fix feels okay but I'd rather create some extra tests around it before merging. I'm just not sure how to properly set up a test beyond a full e2e test, which I'd prefer to skip for this case.

Fixes https://github.com/gatsbyjs/gatsby/issues/25336

Confirmed that at least the given repro case passes with these fixes (https://github.com/stefanprobst/gatsby-issue-link-extension-regression/tree/pr-25109). But not entirely sure whether all edge cases in this part are caught right now.

### CODE REVIEW

File: packages/gatsby/src/schema/resolvers.ts [hunk 1]
```diff
@@ -253,30 +253,29 @@ export function link<TSource, TArgs>(
...
-    const operator = Array.isArray(fieldValue) ? oneOf : equals
-    const runQueryArgs = args as TArgs & { filter: any }
-    runQueryArgs.filter = options.by
-      .split(`.`)
-      .reduceRight((acc, key, i, { length }) => {
-        return {
-          [key]: i === length - 1 ? operator(acc) : acc,
-        }
-      }, fieldValue)
+    const runQueryArgs = args as any
```
Comment (pvdz) [inline packages/gatsby/src/schema/resolvers.ts:261]: Unhappy about this but it's not really changing the status quo. This object is very tricky to type properly due to its semi-dynamic-semi-fixed nature.

Comment (vladar) [reply packages/gatsby/src/schema/resolvers.ts:261]: What's wrong with the previous typing:

```
    const runQueryArgs = args as TArgs & { filter: any }
```

It doesn't give any errors to me with this PR:

**Comment by vladar** [comment]
I've converted it to draft until we have some tests (just to make sure we don't merge it or review prematurely)

File: packages/gatsby/src/schema/resolvers.ts (hunk 1 above)
Comment (pvdz) [reply packages/gatsby/src/schema/resolvers.ts:261]: Not sure at this point. Maybe it was showing errors in an intermediate state but not anymore. I'll revert if so.

Can you help out with the tests?

**Pull request closed without merging**